src/mlpack/methods/reinforcement_learning/replay/hindsight_replay.hpp

```diff
@@ -0,0 +1,423 @@
+   *
+   * @param reward Given reward.
+   * @param nextState Given next state.
+   * @param isEnd Whether next state is terminal state.
+   * @param discount The discount parameter.
+   */
+  void GetNStepInfo(double& reward,
+                    StateType& nextState,
+                    bool& isEnd,
+                    const double& discount)
+  {
+    reward = nStepBuffer.back().reward;
+    nextState = nStepBuffer.back().nextState;
+    isEnd = nStepBuffer.back().isEnd;
+
+    // Should start from the second last transition in buffer.
+    for (int i = nStepBuffer.size() - 2; i >= 0; i--)
+    {
+      bool iE = nStepBuffer[i].isEnd;
+      reward = nStepBuffer[i].reward + discount * reward * (1 - iE);
+      if (iE)
+      {
+        nextState = nStepBuffer[i].nextState;
+        isEnd = iE;
+      }
+    }
+  }
+
+  /**
+   * Sample some experiences.
+   *
+   * @param sampledStates Sampled encoded states.
+   * @param sampledActions Sampled actions.
+   * @param sampledRewards Sampled rewards.
+   * @param sampledNextStates Sampled encoded next states.
+   * @param isTerminal Indicate whether corresponding next state is terminal
+   *        state.
+   */
+  void Sample(arma::mat& sampledStates,
+              std::vector<ActionType>& sampledActions,
+              arma::rowvec& sampledRewards,
+              arma::mat& sampledNextStates,
+              arma::irowvec& isTerminal)
+  {
+    size_t upperBound = full ? capacity : position;
+    arma::uvec sampledIndices = arma::randi<arma::uvec>(
+        batchSize, arma::distr_param(0, upperBound - 1));
+
+    sampledStates = states.cols(sampledIndices);
+    for (size_t t = 0; t < sampledIndices.n_rows; t ++)
+      sampledActions.push_back(actions[sampledIndices[t]]);
+    sampledRewards = rewards.elem(sampledIndices).t();
+    sampledNextStates = nextStates.cols(sampledIndices);
+    isTerminal = this->isTerminal.elem(sampledIndices).t();
+  }
+
+  /**
+   * Get the number of transitions in the memory.
+   *
+   * @return Actual used memory size.
+   */
+  const size_t& Size()
+  {
+    return full ? capacity : position;
+  }
+
+  /**
+   * Update the priorities of transitions and Update the gradients.
+   *
+   * @param target The learned value.
+   * @param sampledActions Agent's sampled action.
+   * @param nextActionValues Agent's next action.
+   * @param gradients The model's gradients.
+   */
+  void Update(arma::mat target,
+              std::vector<ActionType> sampledActions,
+              arma::mat nextActionValues,
+              arma::mat& gradients)
+  {
+    /* Do nothing for hindsight replay. */
+  }
+
+  //! Get the number of steps for n-step agent.
+  const size_t& NSteps() const { return nSteps; }
+
+ private:
+  //! Locally-stored number of examples of each sample.
+  size_t batchSize;
+
+  //! Locally-stored total memory limit.
+  size_t capacity;
+
+  //! Indicate the position to store new transition.
+  size_t position;
```

Comment:
be more explicit

Reply:
I actually took this reference from other replay implementation.

Reply:
We'll be better, please read our guidelines. Variable names should be as descriptive as possible. We won't need commenting if variable names were correct in the first place. Have a [look](https://codegym.cc/groups/posts/369-coding-rules-the-power-of-correct-names-good-and-bad-comments)